---
title: "allele_specific"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Basic Analysis

RNA-seq, snakepipes running:

```{bash basic, echo=FALSE}
##Following vivek protocal, Run all data with same param, trim first 6bp + by quality, and use allele-specific pipeline /data/akhtar/bhardwaj/2017_allelespecific_analysis/allelespecific_RNAseq/02_mapping_counting/runparams.sh, lfc cutoff is 0, fdr cutoff is 0.01
module load snakePipes/2.2.3 && cd /data/akhtar/Mouse2019AlleleSpecific
dir=NPCs9sCa2019RNAseqMouse
dir=ESCs9sCa2019RNAseqMouse #library type is 2
mRNA-seq -i /data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir -o $dir --libraryType 2 -j 20 --DAG --trim --trimmer trimgalore --trimmerOptions '-q 20 --fastqc --trim-n --clip_R1 6' -m allelic-mapping --SNPfile /data/akhtar/suny/Genome/snp_genome/all_CAST_EiJ_SNPs_129S1_SvImJ_reference.based_on_GRCm38.txt --NMaskedIndex /data/akhtar/suny/Genome/snp_genome/star_Nmasked_2.7.4a/Genome --alignerOptions '--outFilterMultimapScoreRange 0 --limitBAMsortRAM 10000000000' mm10 

dir=NPCsBlCa2019RNAseqMouse
dir=NPCsCaBl2019RNAseqMouse
mRNA-seq -i /data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir -o $dir --libraryType 2 -j 20 --DAG --trim --trimmer trimgalore --trimmerOptions '-q 20 --fastqc --trim-n --clip_R1 6' -m allelic-mapping --SNPfile /data/akhtar/suny/Genome/snp_genome_cast/all_SNPs_CAST_EiJ_GRCm38.txt.gz --NMaskedIndex /data/akhtar/suny/Genome/snp_genome_cast/star_Nmasked_2.7.4a/Genome --alignerOptions '--outFilterMultimapScoreRange 0 --limitBAMsortRAM 10000000000' mm10 

##Differential expression analysis
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.control.MSL2KO.all.tsv
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.control.MSL2KO.tsv
mRNA-seq -i /data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir -o $dir --libraryType 2 -m allelic-mapping --SNPfile /data/akhtar/suny/Genome/snp_genome/all_CAST_EiJ_SNPs_129S1_SvImJ_reference.based_on_GRCm38.txt --NMaskedIndex /data/akhtar/suny/Genome/snp_genome/star_Nmasked_2.7.4a/Genome --sampleSheet $samplesheet mm10
#--outFilterMismatchNoverReadLmax 0.02 --tempdir /data/akhtar/Mouse2019AlleleSpecific/

##DEgenes clean up for misssing / duplicated chromosomes
module load R/4.0.3 && cd /data/akhtar/Mouse2019AlleleSpecific/ #source activate R3.6.1 
for dir in NPCs9sCa2019RNAseqMouse NPCsBlCa2019RNAseqMouse NPCsCaBl2019RNAseqMouse ESCs9sCa2019RNAseqMouse ESCsBlCa2019RNAseqMouse ESCsCaBl2019RNAseqMouse
do cd /data/akhtar/Mouse2019AlleleSpecific/$dir/DESeq2_sampleInfo.control.MSL2KO
SlurmEasy -t 20 "Rscript ~/scripts/allele_specific.degenes.cleanup.R /data/akhtar/Mouse2019AlleleSpecific/$dir /data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir DESeq2_sampleInfo.control.MSL2KO"
done

```


TT-seq:
```{bash ttseq,echo = FALSE}
## mapping
cd /data/akhtar/Mouse2019AlleleSpecific
module load snakePipes/2.2.3 && cd /data/akhtar/Mouse2019AlleleSpecific
dir=NPCs9sCa2020RNAseqMouse
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.control.MSL2KO.ttseq.tsv
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.control.MSL2KO.all.ttseq.tsv
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.control.MSL2KO.rnaseq.tsv

mRNA-seq -i /data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir -o $dir --libraryType 1 -m allelic-mapping --SNPfile /data/akhtar/suny/Genome/snp_genome/all_CAST_EiJ_SNPs_129S1_SvImJ_reference.based_on_GRCm38.txt --NMaskedIndex /data/akhtar/suny/Genome/snp_genome/star_Nmasked_2.7.4a/Genome --sampleSheet $samplesheet mm10

```


ATAC-seq:
```{bash basic, echo=FALSE}
## mapping
module load snakePipes/2.2.3 && cd /data/akhtar/Mouse2019AlleleSpecific
dir=NPCs9sCa2019ATACseqMouse
DNA-mapping -i /data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir -o $dir -j 10 --DAG --fastqc --dedup --properPairs --mapq 3 --bwBinSize 25 --trim --trimmer trimgalore --trimmerOptions '-q 20 --fastqc --trim-n' -m allelic-mapping --SNPfile /data/akhtar/suny/Genome/snp_genome/all_CAST_EiJ_SNPs_129S1_SvImJ_reference.based_on_GRCm38.txt --NMaskedIndex /data/akhtar/suny/Genome/snp_genome/bowtie2_Nmasked/Genome.1.bt2 mm10

## ATAC-seq
cd /data/akhtar/Mouse2019AlleleSpecific && conda activate snakepipesDev
dir=NPCs9sCa2019ATACseqMouse
for i in Control MSL2KO1 genome1 genome2; 
do samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.${i}.ATAC.*tsv;
ATAC-seq -d $dir -j 30 --FDR 1 --LFC 0 --CSAWCountsMethod Peak --sampleSheet $samplesheet mm10;
done
#--peakCaller Genrich Control MSL2KO1 MSL2KO2


```

ChIP-seq:

```{bash basic, echo=FALSE}
## mapping
module load snakePipes/2.2.3 && cd /data/akhtar/Mouse2019AlleleSpecific
dir=NPCs9sCa2019ChIPseqMouse
DNA-mapping -i /data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir -o $dir -j 30 --DAG --fastqc --dedup --properPairs --mapq 3 --bwBinSize 25 --trim --trimmer trimgalore --trimmerOptions '-q 20 --fastqc --trim-n' -m allelic-mapping --SNPfile /data/akhtar/suny/Genome/snp_genome/all_CAST_EiJ_SNPs_129S1_SvImJ_reference.based_on_GRCm38.txt --NMaskedIndex /data/akhtar/suny/Genome/snp_genome/bowtie2_Nmasked/Genome.1.bt2 mm10 #--trim --trimmer trimgalore --trimmerOptions '-q 20 --fastqc --trim-n' 

## ChIP-seq
cd /data/akhtar/Mouse2019AlleleSpecific && conda activate snakepipesDev
dir=NPCs9sCa2019ChIPseqMouse
ChIP-seq -d $dir -j 30 --predictChIPDict Chip.yaml mm10 
cat <(sed 's/rep1/rep1.genome1/g' $dir/chip_seq_sample_config.PREDICTED.yaml |sed 's/rep2/rep2.genome1/g' | sed 's/rep3/rep3.genome1/g') <(sed 's/rep1/rep1.genome2/g' $dir/chip_seq_sample_config.PREDICTED.yaml |sed 's/rep2/rep2.genome2/g' | sed 's/rep3/rep3.genome2/g' |sed '1d') > $dir/chip_seq_sample_config.PREDICTED.allelic.yaml && cp $dir/chip_seq_sample_config.PREDICTED.* Fastq/$dir/readme/
ChIP-seq -d $dir -j 30 mm10 Fastq/$dir/readme/chip_seq_sample_config.PREDICTED.yaml

##Allelic Diff analysis
cd /data/akhtar/Mouse2019AlleleSpecific && conda activate snakepipesDev
dir=NPCs9sCa2019ChIPseqMouse
for i in H3K27ac H3K4me3 H4K16ac; 
do
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.genome1.${i}.tsv;
ChIP-seq -d $dir -j 30 --windowSize 100 --FDR 1 --LFC 0 --CSAWCountsMethod Peak --sampleSheet $samplesheet mm10 Fastq/$dir/readme/chip_seq_sample_config.PREDICTED.allelic.yaml 
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.genome2.${i}.tsv;
ChIP-seq -d $dir -j 30 --windowSize 100 --FDR 1 --LFC 0 --CSAWCountsMethod Peak --sampleSheet $samplesheet mm10 Fastq/$dir/readme/chip_seq_sample_config.PREDICTED.allelic.yaml 
done

for i in H3K4me1 H3K27me3 H3K36me3 H3K9me3; 
do
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.genome1.${i}.tsv;
ChIP-seq -d $dir -j 30 --windowSize 500 --FDR 1 --LFC 0 --CSAWCountsMethod Peak --sampleSheet $samplesheet mm10 Fastq/$dir/readme/chip_seq_sample_config.PREDICTED.allelic.yaml 
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.genome2.${i}.tsv;
ChIP-seq -d $dir -j 30 --windowSize 500 --FDR 1 --LFC 0 --CSAWCountsMethod Peak --sampleSheet $samplesheet mm10 Fastq/$dir/readme/chip_seq_sample_config.PREDICTED.allelic.yaml 
done

for i in Control MSL2KO1 MSL2KO2; 
do
samplesheet=/data/akhtar/Mouse2019AlleleSpecific/Fastq/$dir/readme/sampleInfo.${i}.H4K16ac.allelic.tsv;
ChIP-seq -d $dir -j 30 --windowSize 100 --FDR 1 --LFC 0 --sampleSheet $samplesheet mm10 Fastq/$dir/readme/chip_seq_sample_config.PREDICTED.allelic.yaml 
done

```

BS-seq:
```{bash, echo=FALSE}
######################
##merge basic bigwigs
######################
module load WiggleTools/1.2.2
cd /data/akhtar/Mouse2019AlleleSpecific/$dir/bedGraphs_full/
##merge bedgraph to bigwig
for dir in NPCsBlCa2019 NPCsCaBl2019 NPCs9sCa2019
do 
SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs_full/WT*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/Basic/Control.bw"
SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs_full/KO1*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/Basic/MSL2KO1.bw"
SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs_full/KO2*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/Basic/MSL2KO2.bw"
done


######################
##merge allelic bigwigs
######################
module load WiggleTools/1.2.2
cd /data/akhtar/Mouse2019AlleleSpecific/$dir/bedGraphs/
#for i in *.bedGraph.gz; do j=`echo $i |sed 's/bedGraph.gz/sort.bedGraph/g'`; z=`echo $i |sed 's/bedGraph.gz/bw/g'`; SlurmEasy "zcat $i | (head -n 1 ; sort -k1,1 -k2,2n ) > $j && bedGraphToBigWig $j /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes $z"; done
##merge bedgraph to bigwig
#rename 'KO_' 'KO1_' *bw ##if BlCa and CaBl
dir=NPCs9sCa2019 
mat="Svj" && pat="Cast"
dir=NPCsBlCa2019 
mat="C57BL" && pat="CAST"
dir=NPCsCaBl2019
mat="C57BL" && pat="CAST"

SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs/WT*.${mat}*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/AlleleSpecific/Control.genome1.bw"
SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs/WT*.${pat}*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/AlleleSpecific/Control.genome2.bw"
SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs/KO1*.${mat}*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/AlleleSpecific/MSL2KO1.genome1.bw"
SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs/KO1*.${pat}*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/AlleleSpecific/MSL2KO1.genome2.bw"
SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs/KO2*.${mat}*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/AlleleSpecific/MSL2KO2.genome1.bw"
SlurmEasy "wiggletools mean ${dir}BSseqMouse/bedGraphs/KO2*.${pat}*.bg | wigToBigWig stdin /data/repository/organisms/GRCm38_ensembl/genome_fasta/genome.chrom.sizes ${dir}BSseqMouse/Bigwig_merged/AlleleSpecific/MSL2KO2.genome2.bw"
#mkdir -p /data/akhtar/Mouse2019AlleleSpecific/$dir/Bigwig_merged/AlleleSpecific && mv *genome*bw /data/akhtar/Mouse2019AlleleSpecific/$dir/Bigwig_merged/AlleleSpecific

```

## Integrative Analysis: ChIP vs RELACS and RNA-seq vs RELACS
To check if MSL2 allelic binding peak has H4K16ac, H3K9me3 changes upon MSL2 KO.

Expression category analysis, since then we will have 7 categories, and all plots will be based on this
```{bash, echo=FALSE}
upperdir=/data/akhtar/Mouse2019AlleleSpecific/
module load R/4.0.3  #source activate R3.6.1 ##NPCs9sCa2019 pvalue 0.2
for dir in NPCs9sCa2019 NPCsCaBl2019 NPCsBlCa2019 NPCs9sCa2020 NPCs9sCa2019sc ESCs9sCa2019 ESCsBlCa2019 ESCsCaBl2019 ASCsCaBl2019 ASCsBlCa2019 UbiqCaBl2019 Serum9sCa2019
do cd ${upperdir}/${dir}IntegrativeAnalysis && SlurmEasy -t 10 -n DEgenes.number.statistics "Rscript ~/scripts/allele_specific.degenes.category.R ${upperdir}/${dir}RNAseqMouse ${upperdir}/${dir}IntegrativeAnalysis DESeq2_sampleInfo.control.MSL2KO 0.05 2" && cp ${upperdir}/AllCellline2019IntegrativeAnalysis/ExternalPaperSupp/DE.allelic.bi.both.up.bed ${upperdir}/${dir}IntegrativeAnalysis
done
```


That is all currently.
